### DISCUSSION TITLE
Create actor ref before spawning actor

### DISCUSSION THREAD

**Discussion by Sytten** [tqwewe/kameo#97] [Q&A] [answered]
Hi!

Another of my weird questions coming from actix.

In actix an Addr is only a wrapped channel `tx`. This means you can create the ref before spawning an actor.
Here is an example of how that can be used:
```rust
use actix::dev::channel;
use actix::{Actor, Addr, ArbiterHandle, Context};

use crate::MAILBOX_CAPACITY;

pub trait ActixExt
where
    Self: Actor<Context = Context<Self>> + Send + 'static,
{
    fn start_in_arbiter(self, arbiter: &ArbiterHandle) -> Addr<Self> {
        let (tx, rx) = channel::channel(MAILBOX_CAPACITY);

        // create actor
        arbiter.spawn_fn(move || {
            let ctx = Context::with_receiver(rx);
            let fut = ctx.into_future(self);
            actix_rt::spawn(fut);
        });

        Addr::new(tx)
    }
}

impl<A: Actor<Context = Context<Self>> + Send> ActixExt for A {}
```

Essentially that allows you to fix a chicken-and-egg problem when you have two actors that need to reference each other.
In my case it is not two actors, but it is a similar thing (in pseudo code):
```rust
struct ServiceA {
  other_state: Mutex<()>,
  worker_ref: Addr<ServiceAWorker>
}

#[derive(Actor)]
struct ServiceAWorker {
  service: Arc<dyn ServiceADef>
}

impl ServiceA {
  pub fn init() -> Arc<dyn ServiceADef> {
     let svc = Arc::new(ServiceA { .. });
     // spawn worker actor here, give it a clone of svc
     svc
  }
}
```

Any solution to that?

**Comment by tqwewe** [comment]
You might be looking for [`kameo::actor::prepare`](https://docs.rs/kameo/latest/kameo/actor/fn.prepare.html)? That lets you create the ActorRef before the actor has been spawned which I believe is the issue you're referring to?

**Reply by Sytten** [reply]
More or less, I still need to create the actor struct before I can call prepare. I can't craft an ActorRef independently essentially what I am trying to achieve:

1. Create actor ref
2. Create service with a clone of the actor ref
3. Create & start actor with the ref and a clone of the service.

In actix this is possible since the address is only a wrapper around a channel `tx`. I get that it might not be possible here.

**Reply by tqwewe** [reply]
Could `spawn_with` work? I think it would be more ideal if the ActorRef could simply be created as a standalone instance before any of the actor is initiated, similar to actix, but I think `spawn_with` works for now:
```rust
let actor_ref = kameo::actor::spawn_with(|actor_ref| async move {
    // Create service with a clone of the actor ref
    let service = create_service(actor_ref).await;
    
    MyActor { service }
}).await;
```

Annoyingly this doesn't let you do error handling, etc, so I think a better API could be designed, but hopefully it solves your use case?

Additionally, the main branch has `prepare_with`.

**Reply by Sytten** [reply, marked as answer]
Hum I still need to get a clone of my service Arc outside of that spawn so not really. Not too bad, I will add an option on the service side, I was trying to avoid it but it doesn't seem possible at the moment.

So I will do
1. Create service with None actor ref
2. Create actor with service clone
3. Lock service state and set actor ref
